Add retry on rate limiting API responses and timeouts. Fixes #170 and #162.

eesheesh opened the pull request:
Retry on rate-limiting API responses, based on https://developers.google.com/google-apps/calendar/v3/errors and https://tools.ietf.org/html/rfc6585#section-4.

Also retry on network timeout errors.

jonparrott commented on googleapiclient/http.py:
```diff
@@ -56,6 +56,10 @@
 from googleapiclient.model import JsonModel
 from oauth2client import util
 
+try:
```
does `six` not have an alias for this?

eesheesh replied:
@jonparrott not as far as I can tell from [this](http://pythonhosted.org/six/).

jonparrott replied:
Fair enough. Is there any worry about catching socket errors that aren't timeouterrors that we should do different stuff with?

eesheesh replied:
On Windows, in Python 2.7.11, a network timeout looks like this:
```
  File "C:\Python27_11\lib\site-packages\httplib2\__init__.py", line 1609, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File "C:\Python27_11\lib\site-packages\httplib2\__init__.py", line 1351, in _request
    (response, content) = self._conn_request(conn, request_uri, method, body, headers)
  File "C:\Python27_11\lib\site-packages\httplib2\__init__.py", line 1272, in _conn_request
    conn.connect()
  File "C:\Python27_11\lib\site-packages\httplib2\__init__.py", line 928, in connect
    raise socket.error, msg
error: [Errno 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
```

I tried catching `socket.timeout` and it doesn't catch the exception (as expected from this trace), so I believe it's better to retry on a non-retriable error than miss a retriable one - especially one as important as a network timeout.

jonparrott replied:
No worries. See my comment below.

jonparrott commented on tests/test_http.py:
```diff
@@ -704,6 +759,80 @@ def test_no_retry_fails_fast(self):
+
+    try:
+      request.execute()
+      self.fail('Should have raised an exception.')
```
please use `with self.raises()`.

nathanielmanistaatgoogle replied:
I think @jonparrott means `with self.assertRaises(<arguments>)`?

jonparrott replied:
yes, sorry, used to `py.test` and everything melds together in my mind in the morning.

eesheesh replied:
Done.

jonparrott commented on tests/test_http.py:
```diff
@@ -704,6 +759,80 @@ def test_no_retry_fails_fast(self):
+        headers={'content-type': 'application/json'})
+
+    request._rand = lambda: 1.0
+    request._sleep = lambda _: self.fail('sleep should not have been called.')
```
please use a mock and `mock.assert_not_called()`. or `self.assertFalse(mock.called)`.

jonparrott commented on googleapiclient/http.py:
```diff
@@ -91,12 +95,24 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+        reason = data['error']['errors'][0]['reason']
+        logging.warning('Encountered status %d with reason "%s"' %
+                        (resp.status, reason))
+    except Exception:
```
This is too broad. What exactly are you hoping to catch here?

eesheesh replied:
@jonparrott The various cases where parsing failed, e.g. `content` wasn't populated, or `json.loads()` couldn't parse it, or it could parse it but `data['error']['errors'][0]['reason']` is an invalid path... And possibly other things I can't think of.

Is there a way to get a comprehensive list of exceptions that may be thrown? If not, I'd be risking failing the request due to a parsing error in the retry logic, which would be very bad... I'm fairly new to Python so I'd love to learn a better way! :)

jonparrott replied:
iirc `json.loads` only throws `ValueError`. In any case, you should scope the `try: except` as tightly as possible:

```python
if resp and resp.status == 403:
    try:
        data = json.loads(content.decode('utf-8'))
        reason = data['error']['errors'][0]['reason']
    except ValueError, KeyError:
        logging.warning('Invalid JSON content from response')
        reason = content.decode('utf-8')

    logging.warning('Encountered status %d with reason "%s"' %
        (resp.status, reason))
```

eesheesh replied:
Done.

jonparrott commented on googleapiclient/http.py:
```diff
@@ -91,12 +95,24 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+                        (resp.status, reason))
+    except Exception:
+      pass
+    if resp.status < 500 and not ((resp.status == 403 and
```
This is a bit difficult to follow. Could you break it into multiple if statements?

eesheesh replied:
@jonparrott do you mean something like this?
```
    if resp.status < 500:
      if ((resp.status == 403 and
          reason in ['userRateLimitExceeded', 'rateLimitExceeded'])
          or resp.status == 429):
        continue
      else:
        break
```

jonparrott replied:
yes

jonparrott commented:
Finished a review pass. Please let me know when you're ready for another.

eesheesh commented:
Thanks Jon! All set (I hope) - please take another look.

jonparrott commented on googleapiclient/http.py:
```diff
@@ -82,22 +86,45 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+        try:
+          reason = content.decode('utf-8')
+        except UnicodeDecodeError:
+          reason = str(content)
```
I'm unsure on this behavior. If the content can't be decoded as utf-8, it probably doesn't have a useful human-readable string representation. It's probably better to just pass here.

eesheesh replied:
Done.

jonparrott commented:
@eesheesh Thank you for going through these rounds of review with me. This is a critical part of the library and it's important that we get it right.

This is not your fault, but re-reading over the `_retry_request` function I really don't like the way the logic is laid out. We should leave things better than we found them. Do you mind restructuring the function a little bit?

Something like this would be much easier to read:

```python
    resp = None
    content = None

    for retry_num in range(num_retries + 1):

        # Sleep before retrying.
        if retry_num > 0:
            sleep_time = rand() * 2 ** retry_num
            logging.warning(
                'Sleeping %.2f seconds before retry %d of %d for %s: %s %s%s' %
                (sleep_time, retry_num, num_retries, req_type, method, uri,
                ', following status: %d.' % resp.status if resp else ''))
            sleep(sleep_time)

        try:
          resp, content = http.request(uri, method, *args, **kwargs)

        # Retry on ssl errors and socket timeouts.
        except (ssl.SSLError, TimeoutError) as ex:
            # If this is a non-timeout socket error, re-raise it.
            if isinstance(ex, socket.error):
                if ex.errno != errno.ETIMEDOUT:
                    raise

            if retry_num == num_retries:
                raise
            else:
                logging.warning('Encountered exception %s.', repr(ex))
                continue

        # Retry 5xx errors.
        if resp.status >= 500:
            continue

        # Retry 429 errors
        if resp.status == 429:
            continue

        # For 403 errors, we have to check for the `reason` in the response to
        # determine if it is safe to retry.
        if resp.status == 403:
            if not content:
                break

            # Content is in JSON format.
            try:
                data = json.loads(content.decode('utf-8'))
                reason = data['error']['errors'][0]['reason']
            except (ValueError, KeyError):
                logging.warning('Invalid JSON content from response.')
                break

            logging.warning('Encountered status %d with reason "%s"' %
                (resp.status, reason))

            if reason not in ['userRateLimitExceeded', 'rateLimitExceeded']:
                break

    return resp, content

```

(disclaimer: I haven't tested this or anything, but in general this is how I would prefer the logic to flow)

eesheesh commented:
I'll work on these changes now, but just FYI I still won't make a distinction between different `socket.error` types... I did some testing on Python 2.7 across platforms and I don't think it's feasible to identify timeouts reliably. It would be safer to just retry every `socket.error`.

Windows: `socket.error: [Errno 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond`
Linux: `socket.error: [Errno 110] Connection timed out`
Mac: `socket.error: [Errno 60] Operation timed out`

eesheesh commented:
Done, PTAL. :)

jonparrott commented on googleapiclient/http.py:
```diff
@@ -82,22 +88,55 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+        data = json.loads(content.decode('utf-8'))
+        reason = data['error']['errors'][0]['reason']
+      except (UnicodeDecodeError, ValueError, KeyError) as ex:
+        logging.warning('Invalid JSON content from response: %s', repr(ex))
```
I don't think it's necessary/useful to log the exception here.

jonparrott commented:
This is almost good by my standards

> it would be safer to just retry every socket.error

Is `exception.errno` consistent across the platforms (not the value, but it should match the same constant in the `errno` module).

eesheesh commented:
Ok, I'm now checking for the error code. I did leave one case where we retry if the error code isn't found in socket.errno, because reliability  is better than the slight performance gain of not retrying in this corner case.

jonparrott commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      if isinstance(ex, socket.error) and not isinstance(ex, ssl.SSLError):
+        try:
+          # errno's contents differ by platform so we have to match by name.
+          if not (socket.errno.errorcode[ex.errno] in
```
How is this different from

```
import errno

if ex.errno == errno.ETIMEDOUT:

```

eesheesh replied:
The difference is that I need to check for `WSAETIMEDOUT` as well, but it isn't defined when not on Windows. I thought it's more readable and more elegant to check it like that instead of handling `AttributeError`.

jonparrott commented:
Looks mostly good to me. Can you squash your commits. 

@nathanielmanistaatgoogle do you mind taking a final look at this?

(FYI: I'll be out of office until Thursday)

eesheesh commented:
Ok, I *think* I squashed my commits (GitHub noob here), please let me know if I need to do anything else.

Thanks!

nathanielmanistaatgoogle commented:
Not quite squashed - take a look at the way this page says "@eesheesh wants to merge 12 commits into `google:master` from `eesheesh:master`" - there should be just one commit, not twelve.

eesheesh commented:
Sorry, missed a step. PTAL, thanks!

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      sleep_time = rand() * 2 ** retry_num
       logging.warning(
-          'Retry #%d for %s: %s %s%s' % (retry_num, req_type, method, uri,
+          'Sleeping %.2f seconds before retry %d of %d for %s: %s %s%s' %
```
Please don't nest one string interpolation inside another string interpolation?

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
       resp, content = http.request(uri, method, *args, **kwargs)
-    except ssl.SSLError:
+    # Retry on SSL errors and socket timeout errors.
+    except (ssl.SSLError, TimeoutError) as ex:
```
Code of the form
```
except (<exception type 1>, <exception type 2>) as <name>:
  if isinstance(<name>, <exception type>):
    <statements>
```
invites the question of whether it can be of the form
```
except <exception type 1> as <name>:
  <statements>
except <exception type 2> as <possibly other name>:
  <statements>
```
. Does that question apply here? I'm having trouble seeing.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+        try:
+          # errno's contents differ by platform so we have to match by name.
+          if not (socket.errno.errorcode[ex.errno] in
+              ['WSAETIMEDOUT', 'ETIMEDOUT']):
```
Never write a list literal where a tuple literal will do.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
       if retry_num == num_retries:
         raise
       else:
+        logging.warning('Encountered exception %s', repr(ex))
```
What's the difference between `logging.warning('Encountered exception %s', repr(ex))` and `logging.warning('Encountered exception %r', ex)`? Why prefer the former?

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+
+    # For 403 errors, we have to check for the `reason` in the response to
+    # determine if it is safe to retry.
+    if resp.status == 403:
```
`six.moves.http_client.FORBIDDEN` rather than the literal integer value 403, please.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      continue
+
+    # Retry 429 errors.
+    if resp.status == 429:
```
Please define a module-scope constant for this value, probably named `_TOO_MANY_REQUESTS`.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      # If this is a socket.error, only retry if it's a timeout.
+      if isinstance(ex, socket.error) and not isinstance(ex, ssl.SSLError):
+        try:
+          # errno's contents differ by platform so we have to match by name.
```
The depth of code nesting here is pretty fearsome - what opportunity do we have to pull this logic out into a helper function?

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+        logging.warning('Invalid JSON content from response')
+        break
+
+      logging.warning('Encountered status 403 with reason "%s"' % reason)
```
`logging.warning(<format string>, <arguments to format string>)` rather than `logging.warning(<format string> % <arguments to format string>)`.

Fun fact: long ago in a different life I was the one who wrote the patch to pylint that inspects code for this issue.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      logging.warning('Encountered status 403 with reason "%s"' % reason)
+
+      # Only retry rate limit related failures.
+      if reason not in ['userRateLimitExceeded', 'rateLimitExceeded']:
```
Never write a list literal when a tuple literal will do.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented:
(Completed first round of content review.)

eesheesh commented:
I think I addressed all of your comments. I found a way to catch both socket.error and TimeoutError that works on Python 2.7 and 3.4, but fails on Python 2.6 (see Travis results).

It appears that Python 2.6 is obsolete now, do we intend to keep supporting it? If we do, then I can define TimeoutError only when the Python version is 2.6, but it's an ugly hack, and considering that there's no real issue with Python 2.7, I don't think there's any point in filing an upstream bug.

What do you think?

nathanielmanistaatgoogle commented on tests/test_http.py:
```diff
@@ -113,7 +116,46 @@ def request(self, *args, **kwargs):
+          raise ex
+        else:
+          raise TimeoutError()
+
```
Two blank lines between module-scope classes.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -81,23 +84,76 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
   Returns:
     resp, content - Response from the http request (may be HTTP 5xx).
   """
+
```
Drop this inserted blank line. Unless there's some reason for it that I'm missing?

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -56,11 +58,12 @@
+_TOO_MANY_REQUESTS = 429
+
 
 def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
```
With this change this function is now nearly one hundred lines long with seventy lines of body. What options do we have for breaking it up into an assembly of code elements each of which is a manageable size?

eesheesh replied:
I could move the logic to decide what to do based on `resp.status` and `content`, but I think this would be less readable overall. If you think this is better I'm happy to do it.

nathanielmanistaatgoogle replied:
Yes please. A seventy-line method body is fundamentally unreadable.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -81,23 +84,76 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+        data = json.loads(content.decode('utf-8'))
+        reason = data['error']['errors'][0]['reason']
+      except (UnicodeDecodeError, ValueError, KeyError):
+        logging.warning('Invalid JSON content from response')
```
Should we not include the invalid content in what gets logged?

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -81,23 +84,76 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      exception = ex
+    except socket.error as ex:
+      try:
+        # errno's contents differ by platfor,m so we have to match by name.
```
Typo "platfor,m".

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -81,23 +84,76 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+    except socket.error as ex:
+      try:
+        # errno's contents differ by platfor,m so we have to match by name.
+        if not (socket.errno.errorcode[ex.errno] in
```
Prefer `<element> not in <sequence>` to `not <element> in <sequence>`.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -81,23 +84,76 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      except KeyError:
+        pass
+      exception = ex
+    except TimeoutError as ex:
```
In what environment (platform, Python version, and so on) is a `TimeoutError` not also a `socket.error`?

eesheesh replied:
Oh, I didn't know `TimeoutError` is also a `socket.error`! Changed accordingly.

nathanielmanistaatgoogle replied:
Well, I only just found out myself, so don't take my word for it; make your own independent verification.

eesheesh commented:
Thanks for the review, I addressed the comments, please let me know if you want me to break the method down.

eesheesh commented:
Ok, broken into two methods, and rebased the branch to avoid conflicts.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -84,26 +88,90 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+    except socket.error as ex:
+      try:
+        # errno's contents differ by platform, so we have to match by name.
+        if (socket.errno.errorcode[ex.errno] not in
```
Rather than nest a `try` inside an `except`, how about writing `socket.errno.errorcode.get(ex.errno, None)`?

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -84,26 +88,90 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
   return resp, content
 
 
+def _should_retry_response(resp_status, content):
```
Please place this function earlier in the file ("above") any of its use-sites.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -84,26 +88,90 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+    content: The response content body. 
+
+  Returns:
+    True if the response needs to be retried, otherwise False.
```
In the space of ten lines you equate between "should", "justifies", and "needs" - pick one? Probably "should".

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -84,26 +88,90 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      # If we couldn't match it - retry because we can't know for sure.
+      except KeyError:
+        pass
+      exception = ex
```
What's the difference between `exception` and `ex`? Why should this method have both fields? Why doesn't one suffice?

eesheesh replied:
Because of the scope. I'm using `exception` in the logged message. I verified that the caught exception would be out of scope:
```python
>>> test = None
>>> try:
...   raise TimeoutError()
... except Exception as test:
...   pass
... 
>>> test
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'test' is not defined
```

nathanielmanistaatgoogle replied:
Wow. [Today](http://stackoverflow.com/questions/29268892/python-3-exception-deletes-variable-in-enclosing-scope-for-unknown-reason) [I](https://www.python.org/dev/peps/pep-3110/) [learned](https://www.python.org/dev/peps/pep-0344/).

Does `exception` need to be assigned to `None` on both line 92 and line 104?

Since there can't be exactly one field, how about more specific names than `ex` for the shortest-lifetime fields? What would you think of `except ssl.SSLError as ssl_error:` and `except socket.error as socket_error`?

eesheesh replied:
Changes made. Thanks!

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -84,26 +88,90 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
-          ', following status: %d' % resp.status if resp else ''))
+      # Sleep before retrying.
+      sleep_time = rand() * 2 ** retry_num
+      LOGGER.warning('Sleeping %.2f seconds before retry %d of %d for %s: '
```
If you line-wrap after the opening parenthesis, can you avoid the implicit string concatenation? Always avoid implicit string concatenation whenever you can.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented:
@eesheesh: there are still a lot of outstanding code review comments - are you seeing everything I've posted? GitHub code reviews leave a lot to be desired...

eesheesh commented:
@nathanielmanistaatgoogle Sorry for not being more verbose - I replied "Done." now to all of the comments I received and addressed over the last 14 days (since first squishing commits). Please let me know if there's still anything I missed.

Thanks!

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -84,21 +131,37 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+    except socket.error as socket_error:
+      # errno's contents differ by platform, so we have to match by name.
+      if (socket.errno.errorcode.get(socket_error.errno) not in
+          ('WSAETIMEDOUT', 'ETIMEDOUT')):
```
By placing the opening parenthesis of the tuple on the line above you can spare yourself the outer parentheses. I'm under the impression that that's what's considered idiomatic form, but when I try to think about it deeply I just feel embarrassed for the parser.

Someday an auto-formatter will run over this code.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on tests/test_http.py:
```diff
@@ -113,7 +116,45 @@ def request(self, *args, **kwargs):
+      if self.num_errors == 1:
+        raise ssl.SSLError()
+      else:
+        if sys.version_info[0] == 3:
```
`six.PY3` is already imported into this module and is the general way that we determine twoness and threeness throughout this codebase. Any reason why this location should be different?

eesheesh replied:
Done.

nathanielmanistaatgoogle commented on googleapiclient/http.py:
```diff
@@ -84,21 +131,37 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
+      exception = ssl_error
+    except socket.error as socket_error:
+      # errno's contents differ by platform, so we have to match by name.
+      if socket.errno.errorcode.get(socket_error.errno) not in ('WSAETIMEDOUT',
```
Line-wrap _immediately_ after the opening parenthesis. Both elements of the tuple should go on the following line. And while we're dwelling on minutia: add a comma after the last tuple element.

eesheesh replied:
Done.

nathanielmanistaatgoogle commented:
GitHub is not yet showing me squashed commits.

@jonparrott: please rereview? I asked for, and got, many changes. Feel free to tell me which were mistakes on my part.

eesheesh commented:
I haven't squashed commits again yet, waiting to see if you have any additional feedback. Let me know when you want me to squash.

jonparrott commented on googleapiclient/http.py:
```diff
@@ -63,6 +65,51 @@
+    try:
+      data = json.loads(content.decode('utf-8'))
+      reason = data['error']['errors'][0]['reason']
+    except (UnicodeDecodeError, ValueError, KeyError):
```
(no action required): I'd prefer to see a separate catch for KeyError, as it doesn't indicate invalid JSON, just that the JSON wasn't in the format you expected.

jonparrott commented:
This LGTM, and I'm strongly in favor of the separate `should_retry` function - perhaps in the future we could even let users override that.

@eesheesh thank you so much for doing this and sticking with it through our detailed review.

eesheesh commented:
<off topic but...> I updated my branch to resolve conflicts and now I see other people's commits there, and googlebot thinks I don't have a CLA for them.

Can you suggest a safe way to squash the commits on https://github.com/eesheesh/google-api-python-client/commits/master to resolve this?

jonparrott commented:
Looks like some weirdness has occurred with your branch?

Make a backup of your current master, just in case:

   git checkout -b master-backup

Pull upstream (assuming you have google/google-api-python-client as upstream):

   git fetch upstream master

Then maybe try an an interactive rebase to fix your commits on top of master?

   git rebase -i upstream/master

jonparrott commented:
If that fails, I can manually merge this.

eesheesh commented:
That gives me all of these commits:
```
pick 69296f9 Default to octet-stream if mimetype detection fails.
pick 64f0a40 Handle SSL errors with retries
pick 74dfe73 Make compatible with oauth2client v2.0.0
pick 4fffd9c Specify allowed version range for dependencies
pick 88961e5 Fix issue where docs would fail to build if an API is whitelisted.
pick 9e3bc6c Bump version to 1.5.0
pick 7949171 Updated docs
pick e2fb55e Updating CHANGELOG
pick a4b8370 pep8 + print statement alignment
pick ca60f1f Re-adjusted the spacing in the comments.
pick 2a276c6 Make compatible with oauth2client v2.0.0 xsrfutil.py
pick 74a96e7 Fallback to use new discovery uri pattern when the old one fails.
pick 76010c9 move logging into __name__-specific loggers from the root logger
pick 186cb5d Add retry on rate limiting API responses and network timeouts
pick 8848f3e Revert "Add retry on rate limiting API responses and network timeouts"
pick f836a95 Revert "Add retry on rate limiting API responses and network timeouts"
pick c03ff89 Fix merge
pick 8350677 Fix whitespace
```
Should I just squash all of them?

jonparrott commented:
Generally, you can just move all yours to the end, and squash all of yours. Please don't squash someone else's commit into yours.

(If you're still having trouble, I'm happy to just manually merge this).

eesheesh commented:
Ok, I think it's ready now. Let me know if you need any other actions from my side. Thanks!

jonparrott commented:
This is weird, it's still showing 8 commits?

eesheesh commented:
7 of those aren't mine...

I have backups of my changes, is there a way to reset the branch to be exactly like the upstream, and then reapply the changes?

Alternatively, I can try to drop these commits altogether (in the interactive rebase) and see what happens...

jonparrott commented:
I'll just manually merge. Stand by.

jonparrott commented:
I've cherry-picked and merged this as c6425a00e. Thank you so much for your contribution.

eesheesh commented:
Excellent, thanks!
